src/handler/core.rs

```diff
@@ -258,32 +483,93 @@ where
 ///
 /// This function is only used to specify other handlers upon it (see the root
 /// examples).
+///
+/// # Signature
+///
+/// The run-time type signature of this handler is `HandlerSignature::Entry`.
 #[must_use]
 #[track_caller]
-pub fn entry<'a, Input, Output, Descr>() -> Handler<'a, Input, Output, Descr>
+pub fn entry<'a, Output, Descr>() -> Handler<'a, Output, Descr>
 where
-    Input: Send + 'a,
     Output: 'a,
     Descr: HandlerDescription,
 {
-    from_fn_with_description(Descr::entry(), |event, cont| cont(event))
+    from_fn_with_description(Descr::entry(), |event, cont| cont(event), HandlerSignature::Entry)
+}
+
+/// Performs run-time type checking.
+///
+/// If you are using [`DependencyMap`], you should call this function _before_
+/// executing the handler via [`Handler::execute`] or [`Handler::dispatch`]. If
+/// you do _not_ call this function before execution, type checking will be
+/// delayed until the actual execution, which can make things harder to debug.
+///
+/// `sig` can be obtained from [`Handler::sig`].
+///
+/// # Panics
+///
+/// If `container` does not contain all the types that the handler accepts; in
+/// this case, helpful diagnostic information about missing types and code
+/// locations is displayed.
+///
+/// # Examples
+///
+/// ```should_panic
+/// use dptree::prelude::*;
+///
+/// #[derive(Clone)]
+/// struct A;
+/// #[derive(Clone)]
+/// struct B;
+///
+/// // Requires both `A` and `B` to be present in the container.
+/// let handler: Handler<()> = dptree::entry().filter(|_: A| true).endpoint(|_: B| async { () });
+///
+/// // Ok.
+/// dptree::type_check(handler.sig(), &dptree::deps![A, B]);
+///
+/// // Panics with a proper message.
+/// dptree::type_check(handler.sig(), &dptree::deps![A /* Missing `B` */]);
+/// ```
+pub fn type_check(sig: &HandlerSignature, container: &DependencyMap) {
+    match sig {
+        HandlerSignature::Entry => {}
+        HandlerSignature::Other { input_types, output_types: _, obligations } => {
+            let container_types = container
+                .map
+                .iter()
+                .map(|(type_id, dep)| Type { id: *type_id, name: dep.type_name })
+                .collect::<HashSet<_>>();
+
+            if !input_types.is_subset(&container_types) {
```

Comment:
I think doing `input_types.difference(&container_types)` from the start and then  checking if it's empty is a bit nicer

Reply:
How "from the start"? In this function, we compute `container_types` and then immediately check if `input_types` is a subset of `container_types`.